### ISSUE TITLE
Backend pressure

### ISSUE THREAD

**Issue by french-kiwy** [kaltura/nginx-vod-module#1452]
Following the 2 issue #1450 and #1334, I see that our storage has some pressure from small read chunk. 

By trying to change ` vod_initial_read_size` to 16K I can see some problematic behavior in the way to read the remote file.  
If ` vod_initial_read_size` is larger than file header it will first ask for 16k :  

`GET /thr/storage/mp4-320kbps/02/01/0f/file1.mp4 HTTP/1.1
Host: origin.test
User-Agent: VLC/3.0.18 LibVLC/3.0.18
Range: bytes=0-16383
X-Forwarded-For: 10.x.x.60

HTTP/1.1 206 Partial Content
Cache-Control: s-maxage=604800
Content-Type: audio/mp4
Content-Range: bytes 0-16383/3029620
Last-Modified: Thu, 10 May 2018 16:36:55 GMT
Date: Tue, 18 Jul 2023 13:31:42 GMT
Content-Length: 16384`

then it will rerun a request to redownload the same chunk : 
```
GET /thr/storage/mp4-320kbps/02/01/0f/file1.mp4 HTTP/1.1
Host: origin.test
User-Agent: VLC/3.0.18 LibVLC/3.0.18
Cache-Control: no-cache
X-Forwarded-For: 10.x.x.60
Range: bytes=0-16383

HTTP/1.1 206 Partial Content
Cache-Control: s-maxage=604800
Content-Type: audio/mp4
Content-Range: bytes 0-16383/3029620
Last-Modified: Thu, 10 May 2018 16:36:55 GMT
Date: Tue, 18 Jul 2023 13:31:42 GMT
Content-Length: 16384
```

Then it finally start streaming the file :  

```
GET /thr/storage/mp4-320kbps/02/01/0f/file1.mp4 HTTP/1.1
Host: origin.test
User-Agent: VLC/3.0.18 LibVLC/3.0.18
Cache-Control: no-cache
X-Forwarded-For: 10.x.x.60
Range: bytes=13812-94616

HTTP/1.1 206 Partial Content
Cache-Control: s-maxage=604800
Content-Type: audio/mp4
Content-Range: bytes 13812-94616/3029620
Last-Modified: Thu, 10 May 2018 16:36:55 GMT
Date: Tue, 18 Jul 2023 13:31:42 GMT
Content-Length: 80805

**Comment by french-kiwy** [comment]
OK, I will try the request ID trick to see if 2 different request are emitted however those has been run on a test platform where only me and my vlc client were testing so I don't see where a second request would come from. 
Also here's our current (currated) nginx configuration : 

```
user  nginx;
worker_processes  auto;
load_module modules/ngx_http_secure_token_filter_module.so;
load_module modules/ngx_http_vod_module.so;

error_log  /var/log/nginx/error.log warn;
pid        /var/run/nginx.pid;


events {
    worker_connections  1024;
    multi_accept on;
    use epoll;
}

http {
    include       /etc/nginx/mime.types;
    default_type  application/octet-stream;

    log_format  main  '$http_x_forwarded_for $remote_addr - $remote_user [$time_local] "$request" '
                '$status $bytes_sent $request_time "$http_referer" "$http_user_agent" "-" - '
                '"$sent_http_x_kaltura" "$http_host" $pid $sent_http_x_kaltura_session - '
                '$request_length "$sent_http_content_range" "$sent_http_cache_control" $connection ';

    access_log  /var/log/nginx/access.log  main;

    sendfile on;
    tcp_nopush on;
    tcp_nodelay on;

    keepalive_timeout 60;
    keepalive_requests 1000;
    client_header_timeout 20;
    client_body_timeout 20;
    reset_timedout_connection on;
    send_timeout 20;

    gzip on;
    gzip_types application/vnd.apple.mpegurl video/f4m application/dash+xml text/xml;
    upstream http.origin.test.domain.com {
        server http.origin.test.domain.com;
        keepalive 50;
    }

    vod_mode remote;
    vod_upstream_location /proxy_origin;

    vod_metadata_cache metadata_cache 512m;
    vod_mapping_cache mapping_cache 64m;
    vod_response_cache response_cache 64m;
    vod_performance_counters perf_counters;
    open_file_cache max=1000 inactive=5m;
    open_file_cache_valid 2m;
    open_file_cache_min_uses 1;
    open_file_cache_errors on;
    aio on;

    secure_token_ACDN $cdn_token {
        acl                     "https://url.domain.com$secure_token_baseuri*";
        end                     180m;
    }


    server {
        listen 80 default;
        server_name $hostname;
        secure_token $query_string;
        secure_token_types text/xml application/vnd.apple.mpegurl;

        location ~* ^/proxy_origin/(ahlsaes)/ {
                internal;
                rewrite ^/proxy_origin/(?:ahlsaes)/(.*)$ /thr/$1 break;
                proxy_pass http://http.origin.test.domain.com;
                proxy_http_version 1.1;
                proxy_set_header Host "http.origin.test.domain.com";
                proxy_set_header Connection "";
        }

        location /ahlsaes/ {
                secure_token $cdn_token&$query_string;
                vod hls;
                vod_base_url "";
                vod_hls_encryption_method aes-128;
                vod_secret_key "akey";
                vod_bootstrap_segment_durations 2000;
                vod_bootstrap_segment_durations 2000;
                vod_bootstrap_segment_durations 2000;
                vod_manifest_segment_durations_mode accurate;
                vod_segment_duration 4000;
                vod_initial_read_size 16K;

                add_header Access-Control-Allow-Headers "*";
                add_header Access-Control-Expose-Headers "Server,range,Content-Length,Content-Range";
                add_header Access-Control-Allow-Methods "GET, HEAD, OPTIONS";
                add_header Access-Control-Allow-Origin "*";
                add_header Cache-Control "no-store, no-cache, must-revalidate";
        }

        error_page   500 502 503 504  /50x.html;
        location = /50x.html {
                root   html;
        }
    }
}
```

I'll let you know the request ID once I've been able to run more test with request_id enable.

**Comment by french-kiwy** [comment]
Hello, 
Allow me to insists, the issue is still present while I'm the only one using our test server and cache is enable.  
Thank you.

**Comment by erankor** [comment]
Did you run the test I suggested?

**Comment by french-kiwy** [comment]
Yes I no longer have the trace I'll redo them but the first test I run I was already the only one to access those files.
I was doing it on a test env deployed only for my usage.
I'll redo the test and include tcpdump from my VLC client.

**Comment by nuuantoine** [comment]
OK so I've done 2 new tests, one with 8k initial block size here's are the call from hls to backend file to download mp4 file :  
```
GET /url/02/01/0f/file.mp4 HTTP/1.1
Host: origin.domain.com
User-Agent: VLC/3.0.18 LibVLC/3.0.18
Range: bytes=0-8191
X-Forwarded-For: 10.9.8.60

GET /url/02/01/0f/file.mp4 HTTP/1.1
Host: origin.domain.com
User-Agent: VLC/3.0.18 LibVLC/3.0.18
Range: bytes=8192-13795
X-Forwarded-For: 10.9.8.60

GET /url/02/01/0f/file.mp4 HTTP/1.1
Host: origin.domain.com
User-Agent: VLC/3.0.18 LibVLC/3.0.18
Cache-Control: no-cache
X-Forwarded-For: 10.9.8.60
Range: bytes=0-8191

GET /url/02/01/0f/file.mp4 HTTP/1.1
Host: origin.domain.com
User-Agent: VLC/3.0.18 LibVLC/3.0.18
Cache-Control: no-cache
X-Forwarded-For: 10.9.8.60
Range: bytes=8192-13795
```

and here's the same  with 16k : 

```
GET /url/02/01/0f/file.mp4 HTTP/1.1
Host: origin.domain.com
User-Agent: VLC/3.0.18 LibVLC/3.0.18
Range: bytes=0-16383
X-Forwarded-For: 10.9.8.60

GET /url/02/01/0f/file.mp4 HTTP/1.1
Host: origin.domain.com
User-Agent: VLC/3.0.18 LibVLC/3.0.18
Cache-Control: no-cache
X-Forwarded-For: 10.9.8.60
Range: bytes=0-16383

GET /url/02/01/0f/file.mp4 HTTP/1.1
Host: origin.domain.com
User-Agent: VLC/3.0.18 LibVLC/3.0.18
Cache-Control: no-cache
X-Forwarded-For: 10.9.8.60
Range: bytes=13812-94616
```

both test were done after a nginx restart (no cache) and I'm the only one using this server in a test environment. 
In both case kaltura module download file headers twice inducing pressure on  storage in the backend.

Hope those informations help.

**Comment by movy** [comment]
We faced a similar storage-related performance bottleneck, not necessarily caused by vod_module per se, as we have all caches enabled, and until our zfs began re-silvering (and put overlaying GlusterFS on its knees), we did not experience any probs.  However, since this issue is pretty high when googling for similar symptoms, I'll leave the working solution here:

https://www.f5.com/company/blog/nginx/thread-pools-boost-performance-9x

tl;dr: 
add to your vod location: 
```conf
aio threads; # important!
directio       512;
output_buffers 1 128k;
```

(your nginx binary must be built with `--with-threads`). 

This single modification drastically improved i/o load situation and made background resilvering basically unnoticable. Hope it helps someone else.
